bash: fix SIGWINCH issue

If a SIGWINCH arrives while bash is performing redisplay,
multi-line prompts are displayed incorrectly due to the
display code being called recursively.

The patch adds a simple semaphore around the call to
prevent this.

Upstream-Status: Pending

Signed-off-by: Li Wang <li.wang@windriver.com>
---
 lib/readline/display.c  |    4 ++++
 lib/readline/readline.h |    3 ++-
 lib/readline/terminal.c |    2 +-
 3 files changed, 7 insertions(+), 2 deletions(-)

diff --git a/lib/readline/display.c b/lib/readline/display.c
index 2a482ef..3fc4e3b 100644
--- a/lib/readline/display.c
+++ b/lib/readline/display.c
@@ -472,6 +472,8 @@ rl_redisplay ()
   if (!readline_echoing_p)
     return;
 
+  RL_SETSTATE(RL_STATE_SIGWINCH);
+
   if (!rl_display_prompt)
     rl_display_prompt = "";
 
@@ -1190,6 +1192,8 @@ rl_redisplay ()
     else
       visible_wrap_offset = wrap_offset;
   }
+
+  RL_UNSETSTATE (RL_STATE_SIGWINCH);
 }
 
 /* PWP: update_line() is based on finding the middle difference of each
diff --git a/lib/readline/readline.h b/lib/readline/readline.h
index b71bf98..f6e1473 100644
--- a/lib/readline/readline.h
+++ b/lib/readline/readline.h
@@ -797,8 +797,9 @@ extern int rl_inhibit_completion;
 #define RL_STATE_VIMOTION	0x100000	/* reading vi motion arg */
 #define RL_STATE_MULTIKEY	0x200000	/* reading multiple-key command */
 #define RL_STATE_VICMDONCE	0x400000	/* entered vi command mode at least once */
+#define RL_STATE_SIGWINCH	0x800000
 
-#define RL_STATE_DONE		0x800000	/* done; accepted line */
+#define RL_STATE_DONE		0x1000000	/* done; accepted line */
 
 #define RL_SETSTATE(x)		(rl_readline_state |= (x))
 #define RL_UNSETSTATE(x)	(rl_readline_state &= ~(x))
diff --git a/lib/readline/terminal.c b/lib/readline/terminal.c
index 547f6f5..252ee11 100644
--- a/lib/readline/terminal.c
+++ b/lib/readline/terminal.c
@@ -355,7 +355,7 @@ rl_resize_terminal ()
       _rl_get_screen_size (fileno (rl_instream), 1);
       if (CUSTOM_REDISPLAY_FUNC ())
 	rl_forced_update_display ();
-      else
+      else if (RL_ISSTATE(RL_STATE_SIGWINCH) == 0)
 	_rl_redisplay_after_sigwinch ();
     }
 }
-- 
1.7.9.5

